src/sshfs_mount/sshfs_mounts.cpp

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright (C) 2017-2019 Canonical, Ltd.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 3.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include <multipass/exceptions/sshfs_missing_error.h>
+#include <multipass/format.h>
+#include <multipass/logging/log.h>
+#include <multipass/platform.h>
+#include <multipass/ssh/ssh_key_provider.h>
+#include <multipass/sshfs_mount/sshfs_mounts.h>
+#include <multipass/sshfs_server_config.h>
+
+#include <QEventLoop>
+
+namespace mp = multipass;
+namespace mpl = multipass::logging;
+
+namespace
+{
+constexpr auto category = "sshfs-mounts";
+
+template <typename Signal>
+void start_and_block_until(mp::Process* process, Signal signal, std::function<bool(mp::Process* process)> ready_decider)
+{
+    QEventLoop event_loop;
+    auto stop_conn =
+        QObject::connect(process, &mp::Process::finished, [&event_loop](mp::ProcessState) { event_loop.quit(); });
+
+    auto running_conn = QObject::connect(process, signal, [process, ready_decider, &event_loop]() {
+        if (ready_decider(process))
+        {
+            event_loop.quit();
+        }
+    });
+
+    process->start();
+
+    // This blocks and waits either ready_decider to return true, or failure.
+    event_loop.exec();
+
+    QObject::disconnect(stop_conn);
+    QObject::disconnect(running_conn);
+}
+} // namespace
+
+mp::SSHFSMounts::SSHFSMounts(const SSHKeyProvider& key_provider) : key(key_provider.private_key_as_base64())
+{
+}
+
+void mp::SSHFSMounts::start_mount(VirtualMachine* vm, const std::string& source_path, const std::string& target_path,
+                                  const std::unordered_map<int, int>& gid_map,
+                                  const std::unordered_map<int, int>& uid_map)
+{
+    mp::SSHFSServerConfig config;
```

Comment:
You should be able to initialize this directly in the declaration, right?

Reply:
Yes but since it's all strings, I find it too easy to screw up the order of assignments, so prefer the more verbose way. Will change if you want.

Reply:
Ok, at least it's clear what you're doing, so I'm fine with the explicit per member initialization.